crates/wasi-common/src/ctx.rs

```diff
@@ -46,17 +74,29 @@ impl From<OsString> for PendingCString {
 }
 
 impl PendingCString {
-    fn into_string(self) -> Result<String> {
-        match self {
-            Self::Bytes(v) => String::from_utf8(v).map_err(|_| Error::EILSEQ),
-            Self::OsString(s) => s.into_string().map_err(|_| Error::EILSEQ),
-        }
+    fn into_string(self) -> WasiCtxBuilderResult<String> {
+        let res = match self {
+            Self::Bytes(v) => String::from_utf8(v)?,
+            #[cfg(unix)]
+            Self::OsString(s) => {
+                use std::os::unix::ffi::OsStringExt;
+                String::from_utf8(s.into_vec())?
+            }
+            #[cfg(windows)]
+            Self::OsString(s) => {
+                use std::os::windows::ffi::OsStrExt;
+                let bytes: Vec<u16> = s.encode_wide().collect();
+                String::from_utf16(&bytes)?
+            }
+        };
+        Ok(res)
     }
 
-    /// Create a `CString` containing valid UTF-8, or fail with `Error::EILSEQ`.
-    fn into_utf8_cstring(self) -> Result<CString> {
-        self.into_string()
-            .and_then(|s| CString::new(s).map_err(|_| Error::EILSEQ))
+    /// Create a `CString` containing valid UTF-8.
+    fn into_utf8_cstring(self) -> WasiCtxBuilderResult<CString> {
+        let s = self.into_string()?;
+        let s = CString::new(s).map_err(io::Error::from)?;
```

Comment:
This is a super minor nit, but it feels awkward to me to return `Utf8Error` as its own arm of `WasiCtxBuilderResult` while mapping `NulError` to `io::Error`. Also, the `NulError`s we get don't arise from doing any actual I/O. What would you think of adding a new `WasiCtxBuilderError` arm for this?

Reply:
Makes sense! Lemme fix then!